src/Pos/Wallet/Web/Server/Methods.hs

```diff
@@ -330,69 +367,127 @@ getUserProfile = getProfile >>= maybe noProfile pure
 
-sendExtended :: WalletWebMode ssc m => SendActions m -> CPassPhrase -> CAddress -> CAddress -> Coin -> CCurrency -> Text -> Text -> m CTx
+-- TODO [CSM-185]: Send between wallets, not accounts
+sendExtended :: WalletWebMode ssc m => SendActions m -> CPassPhrase -> CAccountAddress -> CAccountAddress -> Coin -> CCurrency -> Text -> Text -> m CTx
 sendExtended sendActions cpassphrase srcCAddr dstCAddr c curr title desc = do
     passphrase <- decodeCPassPhraseOrFail cpassphrase
-    srcAddr <- decodeCAddressOrFail srcCAddr
-    dstAddr <- decodeCAddressOrFail dstCAddr
-    idx <- getAddrIdx srcAddr
-    sks <- getSecretKeys
-    let sk = sks !! idx
+    let srcWCAddr = walletAddrByAccount srcCAddr
+    srcAddr <- decodeCAddressOrFail $ caaAddress srcCAddr
+    dstAddr <- decodeCAddressOrFail $ caaAddress dstCAddr
+    sk      <- getSKByAccAddr passphrase srcCAddr
+    let mainTx = TxOutAux (TxOut dstAddr c) []
+    balance <- getAccountBalance srcCAddr
+    mRems <- if balance < c
+        then return Nothing
+        else do
+            remCAddr <- caAddress <$> newAccount cpassphrase srcWCAddr
+            remAddr  <- decodeCAddressOrFail $ caaAddress remCAddr
+            let remTx = TxOutAux (TxOut remAddr $ balance `unsafeSubCoin` c) []
+            return $ Just (remTx, remCAddr)
     na <- getKnownPeers
     withSafeSigner sk (return passphrase) $ \ss -> do
-        etx <- submitTx sendActions ss na (one $ TxOutAux (TxOut dstAddr c) [])
+        etx <- submitTx sendActions ss na (mainTx :| maybe mempty (one . fst) mRems)
         case etx of
             Left err -> throwM . Internal $ sformat ("Cannot send transaction: "%stext) err
             Right (tx, _, _) -> do
                 logInfo $
-                    sformat ("Successfully sent "%coinF%" from "%ords%" address to "%addressF)
-                    c idx dstAddr
+                    sformat ("Successfully sent "%coinF%" from "%addressF%" address to "%addressF)
+                    c srcAddr dstAddr
                 -- TODO: this should be removed in production
                 let txHash = hash tx
+                whenJust mRems $ \(_, remCAddr) -> do
+                    () <$ addHistoryTx remCAddr curr title desc (THEntry txHash tx True Nothing)
+                    removeAccount srcCAddr
                 () <$ addHistoryTx dstCAddr curr title desc (THEntry txHash tx False Nothing)
                 addHistoryTx srcCAddr curr title desc (THEntry txHash tx True Nothing)
 
 getHistory
     :: forall ssc m.
        (SscHelpersClass ssc, WalletWebMode ssc m)
-    => CAddress -> Maybe Word -> Maybe Word -> m ([CTx], Word)
-getHistory cAddr skip limit = do
-    (minit, cachedTxs) <- transCache <$> getHistoryCache cAddr
+    => CWalletAddress -> Maybe Word -> Maybe Word -> m ([CTx], Word)
+getHistory wAddr skip limit = do
+    cAccAddrs <- getWalletAccAddrsOrThrow wAddr
+    cHistory <- concatForM cAccAddrs $ \cAccAddr -> do
```

Comment:
The best way would be to rewrite history deriving in such way that it supports multiple addresses.